src/signature.rs

```diff
@@ -56,90 +46,27 @@ pub struct Signature {
     ///
     /// This digest is then interpreted as a `Scalar` and reduced into an
     /// element in ℤ/lℤ.
-    pub(crate) s: Scalar,
-}
-
-impl Clone for Signature {
-    fn clone(&self) -> Self {
-        *self
-    }
+    fn s(&self) -> Scalar;
 }
 
-impl Debug for Signature {
-    fn fmt(&self, f: &mut ::core::fmt::Formatter<'_>) -> ::core::fmt::Result {
-        write!(f, "Signature( R: {:?}, s: {:?} )", &self.R, &self.s)
-    }
-}
-
-impl Signature {
-    /// Convert this `Signature` to a byte array.
-    #[inline]
-    pub fn to_bytes(&self) -> [u8; SIGNATURE_LENGTH] {
+impl SignatureExt for Signature {
+    fn from_R_and_s(R: CompressedEdwardsY, s: Scalar) -> Signature {
         let mut signature_bytes: [u8; SIGNATURE_LENGTH] = [0u8; SIGNATURE_LENGTH];
-
-        signature_bytes[..32].copy_from_slice(&self.R.as_bytes()[..]);
-        signature_bytes[32..].copy_from_slice(&self.s.as_bytes()[..]);
-        signature_bytes
+        signature_bytes[..32].copy_from_slice(&R.as_bytes()[..]);
+        signature_bytes[32..].copy_from_slice(&s.as_bytes()[..]);
+        Signature::from_bytes(&signature_bytes[..]).unwrap()
     }
 
-    /// Construct a `Signature` from a slice of bytes.
-    #[inline]
-    pub fn from_bytes(bytes: &[u8]) -> Result<Signature, SignatureError> {
-        if bytes.len() != SIGNATURE_LENGTH {
-            return Err(SignatureError(InternalError::BytesLengthError {
-                name: "Signature",
-                length: SIGNATURE_LENGTH,
-            }));
-        }
+    fn R(&self) -> CompressedEdwardsY {
         let mut lower: [u8; 32] = [0u8; 32];
-        let mut upper: [u8; 32] = [0u8; 32];
-
-        lower.copy_from_slice(&bytes[..32]);
-        upper.copy_from_slice(&bytes[32..]);
-
-        if upper[31] & 224 != 0 {
-            return Err(SignatureError(InternalError::ScalarFormatError));
-        }
-
-        Ok(Signature {
-            R: CompressedEdwardsY(lower),
-            s: Scalar::from_bits(upper),
-        })
+        lower.copy_from_slice(&self.as_ref()[..32]);
+        CompressedEdwardsY(lower)
     }
-}
 
-#[cfg(feature = "serde")]
```

Comment:
The `ed25519` crate provides a `serde` feature, but it uses the bag-of-bytes representation of a signature, and needed some more invasive changes which I didn't complete as part of the initial PR.